contracts/src/ZKEmailProof.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {ERC721URIStorage} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+
+struct ZKEmailProofMetadata {
+    uint256 blueprintId;
+    uint256[] proof;
+    address verifier;
+    string publicOutputs;
+}
+
+/**
+ * @title ZKEmailProof
+ * @notice A soulbound NFT contract that represents valid ZK Email proofs
+ */
+contract ZKEmailProof is ERC721, ERC721URIStorage {
+    using Strings for uint256;
+    using Strings for address;
+
+    error OwnerNotInProof();
+    error CannotTransferSoulboundToken();
+
+    uint256 private _nextTokenId;
+
+    mapping(address owner => ZKEmailProofMetadata metadata)
+        private ownerToMetadata;
+
+    constructor() ERC721("ZKEmailProof", "ZKEP") {}
+
+    /**
+     * @notice Mints a new soulbound NFT representing a ZK email proof
+     * @dev First element of proof must be the recipient address
+     * @param to Address to mint the NFT to
+     * @param blueprintId ID of the blueprint used for the proof
+     * @param proof Proof
+     * @param verifier Address of the verifier contract
+     * @param publicOutputs String of public outputs from the proof
+     */
+    function safeMint(
+        address to,
+        uint256 blueprintId,
+        uint256[] memory proof,
+        address verifier,
+        string memory publicOutputs
+    ) public {
+        // Owner should be committed to in each proof. This prevents
+        // frontrunning safeMint with a valid proof but malicious "to" address
+        if (address(uint160(proof[0])) != to) {
```

Comment:
Wasn't sure on correct format here @Bisht13 - would the address be in the proof like this?

Reply:
this would be publicOutputs[1]